### PR TITLE
Lower dynamic array comparison to __equals template

### PR DESCRIPTION
**Pull request by somzzz**

### CODE REVIEW

File: src/ddmd/expressionsem.d [hunk 1]
```diff
@@ -7859,16 +7859,47 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             }
         }
 
+        Type t1 = exp.e1.type.toBasetype();
+        Type t2 = exp.e2.type.toBasetype();
+
+
+        bool needsDirectEq()
```
Comment (stefan-koch-sociomantic) [inline src/ddmd/expressionsem.d]: could you refactor this to take parameters and not depend on internal state ?
such a refactoring would increase the reusablity.

**Review by andralex — APPROVED** [review]

File: src/ddmd/expressionsem.d [hunk 2]
```diff
@@ -7859,16 +7859,47 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
         }
 
-        if (auto e = typeCombine(exp, sc))
+
+        if (!(t1.ty == Tarray && t2.ty == Tarray && needsDirectEq(t1, t2)))
         {
-            result = e;
-            return;
+            if (auto e = typeCombine(exp, sc))
+                {
```
Comment (stefan-koch-sociomantic) [inline src/ddmd/expressionsem.d]: indentation seems messed up

Comment (somzzz) [reply src/ddmd/expressionsem.d]: fixed

**Review by andralex — APPROVED** [review]
cc @WalterBright

**Comment by JinShil** [comment]
This PR may have introduced the following regression: https://issues.dlang.org/show_bug.cgi?id=18115

**Comment by MartinNowak** [comment]
Looks like it also introduced [18252 – [Reg 2.078] comparison of arrays of associative arrays no longer compiles](https://issues.dlang.org/show_bug.cgi?id=18252).

**Comment by John-Colvin** [comment]
also, this: https://issues.dlang.org/show_bug.cgi?id=18267

**Pull request merged**